### PR TITLE
[SPARK-33592] Fix: Pyspark ML Validator params in estimatorParamMaps may be lost after saving and reloading

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?
Fix: Pyspark ML Validator params in estimatorParamMaps may be lost after saving and reloading

When saving validator estimatorParamMaps, will check all nested stages in tuned estimator to get correct param parent.

Two typical cases to manually test:
~~~python
tokenizer = Tokenizer(inputCol="text", outputCol="words")
hashingTF = HashingTF(inputCol=tokenizer.getOutputCol(), outputCol="features")
lr = LogisticRegression()
pipeline = Pipeline(stages=[tokenizer, hashingTF, lr])

paramGrid = ParamGridBuilder() \
    .addGrid(hashingTF.numFeatures, [10, 100]) \
    .addGrid(lr.maxIter, [100, 200]) \
    .build()
tvs = TrainValidationSplit(estimator=pipeline,
                           estimatorParamMaps=paramGrid,
                           evaluator=MulticlassClassificationEvaluator())

tvs.save(tvsPath)
loadedTvs = TrainValidationSplit.load(tvsPath)

# check `loadedTvs.getEstimatorParamMaps()` restored correctly.
~~~

~~~python
lr = LogisticRegression()
ova = OneVsRest(classifier=lr)
grid = ParamGridBuilder().addGrid(lr.maxIter, [100, 200]).build()
evaluator = MulticlassClassificationEvaluator()
tvs = TrainValidationSplit(estimator=ova, estimatorParamMaps=grid, evaluator=evaluator)

tvs.save(tvsPath)
loadedTvs = TrainValidationSplit.load(tvsPath)

# check `loadedTvs.getEstimatorParamMaps()` restored correctly.
~~~

### Why are the changes needed?
Bug fix.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit test.

### CODE REVIEW

File: python/pyspark/ml/tests/test_tuning.py [hunk 1]
```diff
@@ -401,6 +403,11 @@ def test_save_load_pipeline_estimator(self):
                                   estimatorParamMaps=paramGrid,
                                   evaluator=MulticlassClassificationEvaluator(),
                                   numFolds=2)  # use 3+ folds in practice
+        cvPath = temp_path + "/cv"
+        crossval.save(cvPath)
+        loadedCV = CrossValidator.load(cvPath)
+        self.assertEqual(len(loadedCV.getEstimatorParamMaps()), len(paramGrid))
+        self.assertEqual(loadedCV.getEstimator().uid, crossval.getEstimator().uid)
```
Comment (WeichenXu123) [inline python/pyspark/ml/tests/test_tuning.py:410]: added test.

**Review by zhengruifeng — COMMENTED** [review]
LGTM pending tests

File: python/pyspark/ml/tests/test_tuning.py [hunk 2]
```diff
@@ -73,7 +73,21 @@ def test_addGrid(self):
...
+        self.assertEqual(len(paramMaps1), len(paramMaps2))
+        for paramMap1, paramMap2 in zip(paramMaps1, paramMaps2):
+            self.assertEqual(set(paramMap1.keys()), set(paramMap2.keys()))
+            for param in paramMap1.keys():
+                v1 = paramMap1[param]
+                v2 = paramMap2[param]
+                if isinstance(v1, Params):
+                    self.assertEqual(v1.uid, v2.uid)
+                else:
+                    self.assertEqual(v1, v2)
```
Comment (WeichenXu123) [inline python/pyspark/ml/tests/test_tuning.py:87]: Add strict verification for param maps.

**Review by zhengruifeng — APPROVED** [review]

**Review by srowen — COMMENTED** [review]
Looks OK - test cleanup looks like a win too. Can the refactored logic be reused anywhere else?

**Comment by WeichenXu123** [comment]
> Looks OK - test cleanup looks like a win too. Can the refactored logic be reused anywhere else?

Yes. The MetaAlgorithmReadWrite code will be reused in the PR https://github.com/apache/spark/pull/30471

**Comment by zhengruifeng** [comment]
Merged to master. It seem that this fix can not be directly backported to branch-3.0. @WeichenXu123

**Pull request closed without merging**